EasyPost/Exceptions/API/_Base.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Reflection;
+using EasyPost.Models.API;
+using EasyPost.Utilities;
+using RestSharp;
+
+namespace EasyPost.Exceptions.API
+{
+    public class ApiError : EasyPostError
+    {
+        public readonly string? Code;
+        public readonly List<Error>? Errors;
+        public readonly int? StatusCode;
+
+        /// <summary>
+        ///     Get a formatted error string with expanded details about the error.
+        /// </summary>
+        /// <returns>A formatted error string.</returns>
+        public string PrettyPrintString
+        {
+            get
+            {
+                string errorString = $@"{Code} ({StatusCode}): {Message}";
+                return Errors == null
+                    ? errorString
+                    : Errors.Aggregate(errorString, (current, error) => current + $@"
+                            Field: {error.Field}
+                            Message: {error.Message}
+                    ");
+            }
+        }
+
+        // All constructors for API exceptions are protected, so you cannot directly initialize an instance of the exception class.
+        // Instead, you must use the .FromResponse method to retrieve an instance.
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="ApiError" /> class.
+        /// </summary>
+        /// <param name="errorMessage">Error message string to print to console.</param>
+        /// <param name="statusCode">Optional HTTP status code to store as a property.</param>
+        /// <param name="errorType">Optional error type string to store as a property.</param>
+        /// <param name="errors">Optional list of Error objects to store as a property.</param>
+        protected ApiError(string errorMessage, int? statusCode = null, string? errorType = null, List<Error>? errors = null) : base(errorMessage)
+        {
+            StatusCode = statusCode;
+            Code = errorType;
+            Errors = errors;
+        }
+
+        // great minds think alike: https://github.com/stripe/stripe-dotnet/blob/6b9513d3b938d265c7607db919ad2c536ab578c3/src/Stripe.net/Infrastructure/Public/StripeClient.cs#L171
+
+        /// <summary>
+        ///     Parse a RestResponse response object and return an instance of the appropriate exception class.
+        ///     Do not pass a non-error response to this method.
+        /// </summary>
+        /// <param name="response">RestResponse response to parse</param>
+        /// <raises>EasyPostError if a non-400/500 error code is extracted from the response.</raises>
+        /// <returns>An instance of an HttpError-inherited exception.</returns>
+        internal static ApiError FromResponse(RestResponse response)
```

Comment:
Thought: If we are so concerned about not using this for success, should we instead rename it to `FromErrorResponse` or something similar to more clearly call out what this is for?

Reply:
Resolved